### PR TITLE
Review examples for dynamic optimal control; fix iLQR

### PR DESCRIPTION
**Pull request by wxmerkt**
- Makes iLQR faster by using a backtracking line-search and stopping early (like DDP solvers)
- Updates some of the examples to work again, and faster

Notes:
- The iLQG examples do not work
- The Valkyrie examples do not make sense
- The other dynamic examples do not work as of now

### CODE REVIEW

File: exotations/solvers/exotica_ddp_solver/src/abstract_ddp_solver.cpp [hunk 1]
```diff
@@ -184,16 +184,16 @@ void AbstractDDPSolver::Solve(Eigen::MatrixXd& solution)
             }
             else
             {
-                if (debug_) HIGHLIGHT("Cost improved, decrease regularization");
-                DecreaseRegularization();
+                // if (debug_) HIGHLIGHT("Cost improved, decrease regularization");
+                if (lambda_ > 1e-12) DecreaseRegularization();
```
Comment (traikodinev) [inline exotations/solvers/exotica_ddp_solver/src/abstract_ddp_solver.cpp]: Maybe export that `1e-12` as a constant somewhere?

File: exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp [hunk 2]
```diff
@@ -42,32 +42,32 @@ void ILQRSolver::SpecifyProblem(PlanningProblemPtr pointer)
...
+    const Eigen::MatrixXd& Qf = prob_->get_Qf();
+    const Eigen::MatrixXd R = dt * prob_->get_R();
+    const Eigen::MatrixXd& X_star = prob_->get_X_star();
 
     // eq. 18
     vk_gains_[T - 1] = Qf * dynamics_solver_->StateDelta(prob_->get_X(T - 1), X_star.col(T - 1));
-    vk_gains_[T - 1] = vk_gains_[T - 1].unaryExpr([min_clamp_, max_clamp_](double x) -> double {
-        return std::min(std::max(x, min_clamp_), max_clamp_);
-    });
+    // vk_gains_[T - 1] = vk_gains_[T - 1].unaryExpr([min_clamp_, max_clamp_](double x) -> double {
```
Comment (traikodinev) [inline exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp]: This clamping was from Emo's code here: https://homes.cs.washington.edu/~todorov/software/iLQG.zip

Seems sensible... why did you remove it?

File: exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp [hunk 3]
```diff
@@ -176,88 +189,119 @@ void ILQRSolver::Solve(Eigen::MatrixXd& solution)
...
-            global_best_U = new_U;
-            best_ref_x_ = ref_x;
-            best_ref_u_ = ref_u;
+            // Function tolerance patience check
+            if (parameters_.FunctionTolerancePatience > 0)
+            {
+                if (iteration - last_best_iteration > parameters_.FunctionTolerancePatience)
+                {
+                    if (debug_) HIGHLIGHT_NAMED("ILQRSolver", "Early stopping criterion reached (" << cost << " < " << cost_prev << "). Time: " << planning_timer.GetDuration());
+                    prob_->termination_criterion = TerminationCriterion::FunctionTolerance;
```
Comment (traikodinev) [inline exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp]: I'm not sure there shouldn't be a different stopping criteria -- `EarlyStopping`.

File: exotations/solvers/exotica_ddp_solver/src/abstract_ddp_solver.cpp (hunk 1 above)
Comment (traikodinev) [inline exotations/solvers/exotica_ddp_solver/src/abstract_ddp_solver.cpp]: Maybe export that 1e-12 as a constant somewhere?

File: exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp (hunk 2 above)
Comment (traikodinev) [inline exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp]: This clamping was from Emo's code here: https://homes.cs.washington.edu/~todorov/software/iLQG.zip

Seems sensible... why did you remove it?

File: exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp [hunk 4]
```diff
@@ -176,88 +189,119 @@ void ILQRSolver::Solve(Eigen::MatrixXd& solution)
...
+                    break;
+                }
+            }
+            else
+            {
+                if (debug_) HIGHLIGHT_NAMED("ILQRSolver", "Function tolerance reached (" << cost << " < " << cost_prev << "). Time: " << planning_timer.GetDuration());
+                prob_->termination_criterion = TerminationCriterion::FunctionTolerance;
```
Comment (traikodinev) [inline exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp]: I'm not sure there shouldn't be a different stopping criteria -- `EarlyStopping`

File: exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp (hunk 2 above)
Comment (wxmerkt) [reply exotations/solvers/exotica_ilqr_solver/src/ilqr_solver.cpp]: It seemed unnecessary. In that case, happy to add it back.
